pkg/package-server/provider/registry.go

```diff
@@ -0,0 +1,533 @@
+	return prov
+}
+
+func (p *RegistryProvider) getSubscribers(namespace string, eventType pkgEventType) map[chan<- v1alpha1.PackageManifest]subInfo {
+	subscribers := make(map[chan<- v1alpha1.PackageManifest]subInfo)
+
+	// If the namespace is global, return all subscribers of the given eventType.
+	if namespace == p.globalNamespace {
+		for key, subs := range p.subscribers {
+			if key.eventType == eventType {
+				for sub, info := range subs {
+					subscribers[sub] = info
+				}
+			}
+		}
+		return subscribers
+	}
+
+	key := subKey{
+		namespace: namespace,
+		eventType: eventType,
+	}
+	subscribers = p.subscribers[key]
+
+	return subscribers
+}
+
+func (p *RegistryProvider) publishEvents(ctx context.Context, logger *logrus.Entry, pkgs []v1alpha1.PackageManifest, namespace string, eventType pkgEventType) {
+	logger = logger.WithField("eventType", eventType)
+
+	p.smu.RLock()
+	defer p.smu.RUnlock()
+
+	// Get the set of subscribers to send events to.
+	subscribers := p.getSubscribers(namespace, eventType)
+	if len(subscribers) == 0 {
+		logger.Warn("no subscribers found")
+		return
+	}
+
+	logger.WithField("subscribers", len(subscribers)).Info("found subscribers")
+
+	var subGroup sync.WaitGroup
+	subGroup.Add(len(subscribers) * len(pkgs))
+	for sub, info := range subscribers {
+		subCtx, cancel := context.WithTimeout(info.ctx, defaultEventTimeout)
+		defer cancel()
+
+		// Signal that the subscriber's channel is being written to and is not safe to close.
+		info.wg.Add(1)
+		defer info.wg.Done()
+
+		for _, pkg := range pkgs {
+			go func(c context.Context, s chan<- v1alpha1.PackageManifest, p v1alpha1.PackageManifest) {
+				defer subGroup.Done()
+
+				select {
+				case s <- p:
+					logger.Info("event written")
+					return
+				case <-ctx.Done():
+					logger.WithField("err", ctx.Err()).Warn("publish context closed while sending event")
+					return
+				case <-c.Done():
+					logger.WithField("err", c.Err()).Warn("subscription context closed while sending event")
+					return
+				}
+			}(subCtx, sub, pkg)
+		}
+	}
+
+	subGroup.Wait()
+}
+
+func (p *RegistryProvider) catalogSourceAdded(obj interface{}) {
+	catsrc, ok := obj.(*operatorsv1alpha1.CatalogSource)
+	if !ok {
+		logrus.Debugf("casting catalog source failed: wrong type: %#v", obj)
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"action":    "CatalogSource Added",
+		"name":      catsrc.GetName(),
+		"namespace": catsrc.GetNamespace(),
+	})
+
+	conn, err := grpc.Dial(catsrc.Status.RegistryServiceStatus.Address(), grpc.WithInsecure())
+	if err != nil {
+		logger.Errorf("could not connect to registry service: %s", err)
+		return
+	}
+
+	func() {
+		p.cmu.Lock()
+		defer p.cmu.Unlock()
+		p.clients = append(p.clients, newRegistryClient(catsrc, conn))
+	}()
+
+	pkgList, err := p.List(catsrc.GetNamespace())
+	if err != nil {
+		logger.Warn("could not list packages to generate add events")
+		return
+	}
+	pkgs := pkgList.Items
+
+	if catsrc.GetNamespace() != p.globalNamespace {
```

Comment:
Do we need this condition? Shouldn't we just always send the global packages?

Reply:
yes - but the way the code is organized here, if the request namespace  _is_ the global one,  you don't want to end up with duplicates. I decided to add this here instead of at the List(...) level because I thought we were playing with the idea of removing the concept of global packages altogether.